File: back-end/src/auth/auth.service.ts
import {
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { UsersService } from '../users/users.service';
import { RegisterDto } from './dto/register.dto';
import { LoginDto } from './dto/login.dto';
import { JwtPayload } from '../common/interfaces/jwt-payload.interface';
import { User } from '../users/user.entity';

@Injectable()
export class AuthService {
  constructor(
    private readonly usersService: UsersService,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
  ) {}

  async register(dto: RegisterDto) {
    const user = await this.usersService.createUser(dto);
    return this.buildAuthResponse(user);
  }

  async login(dto: LoginDto) {
    const user = await this.usersService.validateUser(dto.email, dto.password);
    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }
    return this.buildAuthResponse(user);
  }

  private buildAuthResponse(user: User) {
    const payload: JwtPayload = {
      sub: user.id,
      email: user.email,
      name: user.name,
    };
    const accessToken = this.jwtService.sign(payload, {
      secret: this.configService.get('JWT_SECRET', 'super-secret'),
      expiresIn: '24h',
    });
    const { passwordHash, ...safeUser } = user;
    return { accessToken, user: safeUser };
  }
}

File: back-end/app/services/auth_service.py
from sqlalchemy.orm import Session
from datetime import timedelta
from app.services.user_service import UserService
from app.schemas.auth import Register, Login, AuthResponse, TokenPayload
from app.schemas.user import UserResponse
from app.core.security import create_access_token
from app.core.config import settings


class AuthService:
    @staticmethod
    def register(db: Session, dto: Register) -> AuthResponse:
        """Register a new user"""
        user = UserService.create_user(db, dto)
        return AuthService._build_auth_response(user)

    @staticmethod
    def login(db: Session, dto: Login) -> AuthResponse:
        """Login user"""
        user = UserService.validate_user(db, dto.email, dto.password)
        if not user:
            from fastapi import HTTPException, status
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid credentials",
            )
        return AuthService._build_auth_response(user)

    @staticmethod
    def _build_auth_response(user) -> AuthResponse:
        """Build authentication response with JWT token"""
        payload = TokenPayload(
            sub=user.id,
            email=user.email,
            name=user.name,
        )
        access_token = create_access_token(
            payload.model_dump(),
            expires_delta=timedelta(hours=settings.JWT_EXPIRATION_HOURS),
        )
        return AuthResponse(
            access_token=access_token,
            user=UserResponse.model_validate(user),
        )

